provider/maas/environ.go

```diff
@@ -1266,70 +1271,53 @@ func deploymentStatusCall(nodes gomaasapi.MAASObject, ids ...instance.Id) (gomaa
 }
 
 type selectNodeArgs struct {
-	AvailabilityZones []string
-	NodeName          string
-	SystemId          string
-	Constraints       constraints.Value
-	Interfaces        []interfaceBinding
-	Volumes           []volumeInfo
+	AvailabilityZone string
+	NodeName         string
+	SystemId         string
+	Constraints      constraints.Value
+	Interfaces       []interfaceBinding
+	Volumes          []volumeInfo
 }
 
 func (environ *maasEnviron) selectNode(args selectNodeArgs) (*gomaasapi.MAASObject, error) {
-	var err error
-	var node gomaasapi.MAASObject
-
-	for i, zoneName := range args.AvailabilityZones {
-		node, err = environ.acquireNode(
-			args.NodeName,
-			zoneName,
-			args.SystemId,
-			args.Constraints,
-			args.Interfaces,
-			args.Volumes,
-		)
-
-		if err, ok := errors.Cause(err).(gomaasapi.ServerError); ok && err.StatusCode == http.StatusConflict {
-			if i+1 < len(args.AvailabilityZones) {
-				logger.Infof("could not acquire a node in zone %q, trying another zone", zoneName)
-				return nil, environs.ErrAvailabilityZoneFailed
-			}
-		}
-		if err != nil {
-			return nil, errors.Errorf("cannot run instances: %v", err)
-		}
-		// Since a return at the end of the function is required
-		// just break here.
-		break
+	node, err := environ.acquireNode(
+		args.NodeName,
+		args.AvailabilityZone,
+		args.SystemId,
+		args.Constraints,
+		args.Interfaces,
+		args.Volumes,
+	)
+	if args.AvailabilityZone != "" && isConflictError(err) {
+		logger.Debugf("could not acquire a node in zone %q", args.AvailabilityZone)
+		return nil, environs.ErrAvailabilityZoneFailed
```

Comment:
Is it ok to upgrade any error to AZ error? Why do we need AZ error in the first place. Shouldn't any failure to provision be tried in another zone?
The old code only promoted to AZ error if it knew (with an off-by-one bug) that there was another AZ that could be tried.
Do we have any way to do something similar here?

Reply:
Yeah, I don't like that particular part. It's not possible at the moment, because the provisioner relies on ErrAvailabilityZoneFailed to decide whether or not to attempt another zone. i.e. anything that's not ErrAvailabilityZoneFailed is expected to be a zone-independent failure, and is terminal.

It's not for this PR, but I think we should revisit this. One idea I have is to introduce a couple of error interfaces to the `environs` package, in the vein of net.Error, that both the provisioner and bootstrap code can rely upon:

```go
type StartInstanceError interface {
    error

    // Permanent reports whether or not the error is permanent.
    // Permanent errors will not be retried by the provisioner.
    Permanent() bool
}

type AvailabilityZoneError interface {
    error

    // AvailabilityZoneIndependent reports whether or not the
    // error is related to a specific availability zone.
    AvailabilityZoneIndependent() bool
}
```

The provisioner would assume that all errors are transient unless they implement StartInsanceError and Permanent returns true. Similarly, it would assume that all errors are zone-specific and attempt the next zone, unless they implement AvailabilityZoneError and AvailabilityZoneIndependent returns true.